### PR TITLE
Improve incremental merge on partitioned models

### PR DESCRIPTION
**Pull request by jacobus-herman**
### Description

At Otrium I have been seeing very poor incremental merge performance on large tables that have partitions. For some reason, Databricks doesn't do partition pruning on the destination table when merging a set of data that only touches a few partitions. I fixed this in our production dbt project by:

1. Executing a simple query that gets the min and max value of each partition column of the update/source dataset.
2. Adding a filter on the destination table's partition columns using the min and max values previously retrieved.

The code doesn't change the merge statement when no `partition_by` or `liquid_clustered_by` is specified.

I tested this code on Otrium's Databricks for AWS workspace using a SQL Classic Warehouse (preview channel - v 2023.50).
I didn't add tests yet, as I am not sure how thoroughly you want this tested but I'm happy to add some :)

### Checklist

- [yes] I have run this code in development and it appears to resolve the stated issue
- [yes] This PR includes tests, or tests are not required/relevant for this PR
- [no] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

**Comment by benc-db** [comment]
@jacobus-herman I think we would like to see at least macro unit tests that validate that we generate the expected code.  I'm going to make a new branch to run the existing E2E tests to see if this doesn't disrupt any existing behavior.

File: tests/unit/macros/materializations/incremental/test_convert_to_list_macro.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    @pytest.fixture(scope="class")
+    def macro_folders_to_load(self) -> list:
+        return ["macros/materializations/incremental"]
+
+    @pytest.fixture(scope="class")
+    def template_name(self) -> str:
+        return "convert_to_list.sql"
+
+    @pytest.mark.parametrize(
+        "func_input,expected", [(["a"], ["a"]), (None, []), ([], [])]
```
Comment (benc-db) [inline tests/unit/macros/materializations/incremental/test_convert_to_list_macro.py:16]: Recommend adding a string -> list test case.

Comment (jacobus-herman) [reply tests/unit/macros/materializations/incremental/test_convert_to_list_macro.py]: Sure, I added that one.

**Comment by benc-db** [comment]
@jacobus-herman Sorry for the delay, I've been squashing a bunch of bugs this week.  Will run the integration tests one last time and prepare for merge either today or Monday.  Thanks for your patience and contribution.

**Comment by jacobus-herman** [comment]
Hey @benc-db  ! Is there something more I can do or is it just a bit busy to merge?

**Comment by benc-db** [comment]
> Hey @benc-db ! Is there something more I can do or is it just a bit busy to merge?

It's just a little chaotic on my side at the moment, apologies for the delay.

**Comment by jacobus-herman** [comment]
> > Hey @benc-db ! Is there something more I can do or is it just a bit busy to merge?
> 
> It's just a little chaotic on my side at the moment, apologies for the delay.

@benc-db this is no longer needed. I was able to do some changes to our underlying tables which resulted in a greater speedup without this can than with it. So I think it's better to leave the functionality as it :)

**Pull request closed without merging**